Source/InvocationCollection.cs

```diff
@@ -78,49 +78,120 @@ public int Count
 
 		public void Add(Invocation invocation)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Add(invocation);
+				if (this.count == this.capacity)
+				{
+					var targetCapacity = this.capacity == 0 ? 4 : (this.capacity * 2);
+					Array.Resize(ref this.invocations, targetCapacity);
+					this.capacity = targetCapacity;
+				}
+
+				this.invocations[this.count] = invocation;
+				this.count++;
 			}
 		}
 
 		public bool Any()
 		{
-			return this.invocations.Count > 0;
+			lock (this.invocationsLock)
+			{
+				return this.count > 0;
+			}
 		}
 
 		public void Clear()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Clear();
+				// Replace the collection so readers with a reference to the old collection aren't interrupted
+				this.invocations = null;
+				this.count = 0;
+				this.capacity = 0;
 			}
 		}
 
 		public Invocation[] ToArray()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				return this.invocations.Take(this.count).ToArray();
 			}
 		}
 
 		public Invocation[] ToArray(Func<Invocation, bool> predicate)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.Where(predicate).ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				return this.invocations.Take(this.count).Where(predicate).ToArray();
 			}
 		}
 
 		public IEnumerator<IReadOnlyInvocation> GetEnumerator()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.ToList().GetEnumerator();
+				return new Enumerator(this.invocations, this.count);
```

Comment:
I'm wondering, why not the LINQ range-based approach you suggested earlier?

Reply:
`Range` creates an enumerator which operates on the underlying collection. If the collection is modified on another thread whilst working with this enumerator, at *best* you'll get an `InvalidOperationException` and at worse will end up in an unknown state. Previously, copying was used to prevent modification by another thread. This custom `Enumerator` instead takes advantage of append-only nature of the collection and allows enumeration over the known state of the collection even when the collection itself is appending new items to the collection.

Reply:
OK, fair enough, that wouldn't work. But see my comment above regarding a plain indexing `for` loop.